BACKPORT - revert to older code for creating sysfs driver groups

This reverts commit 23b9c1ab5baf368a32b7242bf110ef1f48700d04 in 2.6.25
    
Signed-off-by: Ralph Campbell <ralph.campbell@qlogic.com>

---
 drivers/infiniband/hw/ipath/ipath_driver.c |   17 +++++++++++++----
 drivers/infiniband/hw/ipath/ipath_kernel.h |    3 ++-
 drivers/infiniband/hw/ipath/ipath_sysfs.c  |   19 ++++++++++++++-----
 3 files changed, 29 insertions(+), 10 deletions(-)

Index: ofed_kernel/drivers/infiniband/hw/ipath/ipath_driver.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/ipath/ipath_driver.c
+++ ofed_kernel/drivers/infiniband/hw/ipath/ipath_driver.c
@@ -151,9 +151,6 @@ static struct pci_driver ipath_driver = 
 	.probe = ipath_init_one,
 	.remove = __devexit_p(ipath_remove_one),
 	.id_table = ipath_pci_tbl,
-	.driver = {
-		.groups = ipath_driver_attr_groups,
-	},
 };
 
 static inline void read_bars(struct ipath_devdata *dd, struct pci_dev *dev,
@@ -2544,15 +2541,25 @@ static int __init infinipath_init(void)
 		goto bail_unit;
 	}
 
+	ret = ipath_driver_create_group(&ipath_driver.driver);
+	if (ret < 0) {
+		printk(KERN_ERR IPATH_DRV_NAME ": Unable to create driver "
+		       "sysfs entries: error %d\n", -ret);
+		goto bail_pci;
+	}
+
 	ret = ipath_init_ipathfs();
 	if (ret < 0) {
 		printk(KERN_ERR IPATH_DRV_NAME ": Unable to create "
 		       "ipathfs: error %d\n", -ret);
-		goto bail_pci;
+		goto bail_group;
 	}
 
 	goto bail;
 
+bail_group:
+	ipath_driver_remove_group(&ipath_driver.driver);
+
 bail_pci:
 	pci_unregister_driver(&ipath_driver);
 
@@ -2567,6 +2574,8 @@ static void __exit infinipath_cleanup(vo
 {
 	ipath_exit_ipathfs();
 
+	ipath_driver_remove_group(&ipath_driver.driver);
+
 	ipath_cdbg(VERBOSE, "Unregistering pci driver\n");
 	pci_unregister_driver(&ipath_driver);
 
Index: ofed_kernel/drivers/infiniband/hw/ipath/ipath_kernel.h
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/ipath/ipath_kernel.h
+++ ofed_kernel/drivers/infiniband/hw/ipath/ipath_kernel.h
@@ -1271,7 +1271,8 @@ struct device_driver;
 
 extern const char ib_ipath_version[];
 
-extern struct attribute_group *ipath_driver_attr_groups[];
+int ipath_driver_create_group(struct device_driver *);
+void ipath_driver_remove_group(struct device_driver *);
 
 int ipath_device_create_group(struct device *, struct ipath_devdata *);
 void ipath_device_remove_group(struct device *, struct ipath_devdata *);
Index: ofed_kernel/drivers/infiniband/hw/ipath/ipath_sysfs.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/ipath/ipath_sysfs.c
+++ ofed_kernel/drivers/infiniband/hw/ipath/ipath_sysfs.c
@@ -1069,11 +1069,6 @@ static ssize_t show_tempsense(struct dev
 	return ret;
 }
 
-struct attribute_group *ipath_driver_attr_groups[] = {
-	&driver_attr_group,
-	NULL,
-};
-
 static DEVICE_ATTR(guid, S_IWUSR | S_IRUGO, show_guid, store_guid);
 static DEVICE_ATTR(lmc, S_IWUSR | S_IRUGO, show_lmc, store_lmc);
 static DEVICE_ATTR(lid, S_IWUSR | S_IRUGO, show_lid, store_lid);
@@ -1181,6 +1176,20 @@ int ipath_expose_reset(struct device *de
 	return ret;
 }
 
+int ipath_driver_create_group(struct device_driver *drv)
+{
+	int ret;
+
+	ret = sysfs_create_group(&drv->kobj, &driver_attr_group);
+
+	return ret;
+}
+
+void ipath_driver_remove_group(struct device_driver *drv)
+{
+	sysfs_remove_group(&drv->kobj, &driver_attr_group);
+}
+
 int ipath_device_create_group(struct device *dev, struct ipath_devdata *dd)
 {
 	int ret;
